smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGenerator.kt

```diff
@@ -89,27 +92,24 @@ abstract class HttpProtocolClientGenerator(
         writer.dependencies.addAll(KotlinDependency.HTTP.dependencies)
     }
 
-    //  defaults to Ktor since it's the only available engine in smithy-kotlin runtime
     /**
-     * The client engine to default to when one is not given in config. This type *MUST* be default constructable
-     * or else you need to override [renderInit] and construct it manually
+     * Render the class initialization block, the base behavior of which is to accept and configure the HTTP client engine.
+     * Use [renderAdditionalInit] to add logic within this block.
      */
-    protected open val defaultHttpClientEngineSymbol: Symbol = buildSymbol {
-        name = "DefaultHttpEngine"
-        namespace(KotlinDependency.DEFAULT_HTTP_ENGINE)
-    }
+    private fun renderInit(writer: KotlinWriter) {
+        writer.withBlock("init {", "}") {
+            withBlock("if (config.httpClientEngine is #T) {", "}", RuntimeTypes.Http.Engine.ManagedHttpClientEngine) {
+                write("config.httpClientEngine.share()")
+            }
+            write("client = #T(config.httpClientEngine)", RuntimeTypes.Http.SdkHttpClientFn)
 
-    /**
-     * Render the class initialization block. By default this configures the HTTP client
-     */
-    protected open fun renderInit(writer: KotlinWriter) {
-        writer.addImport(defaultHttpClientEngineSymbol)
-        writer.openBlock("init {", "}") {
-            writer.write("val httpClientEngine = config.httpClientEngine ?: #T()", defaultHttpClientEngineSymbol)
-            writer.write("client = sdkHttpClient(httpClientEngine, manageEngine = config.httpClientEngine == null)")
+            write("")
+            renderAdditionalInit(writer)
         }
     }
```

Comment:
**Question**: I only see HTTP engine being shared here. Will credentials provider sharing come in a future PR?